Add initial Blackboard support

patdhlk opened the pull request:
This pull request adds a new example demonstrating the Blackboard communication pattern in the C#/.NET bindings for iceoryx2. It introduces a complete `Blackboard` example project, updates documentation to include the new pattern, and makes minor improvements to formatting and naming for clarity and consistency.

**Major new feature: Blackboard pattern support and example**
- Adds a full example project (`examples/Blackboard`) showcasing the Blackboard key-value store pattern, including both creator (writer) and opener (reader) roles, with sample code for sensor data sharing. [[1]](diffhunk://#diff-3c176fa8aa8afe2c6c5528498fe84c40472220071d9703d33481ec85074c6d3eR1-R37) [[2]](diffhunk://#diff-73240cf76b6c8604543951acac7304ce2097ec4e2d292174e7b59c07e05ff5bbR1-R279)
- Updates the solution file (`Iceoryx2.sln`) to include the new `Blackboard` example project and its build configurations. [[1]](diffhunk://#diff-f97b528be07ea2a25c936948697e54a3fd1a2a92b02178888aae827b0334ee78R48-R49) [[2]](diffhunk://#diff-f97b528be07ea2a25c936948697e54a3fd1a2a92b02178888aae827b0334ee78R288-R299) [[3]](diffhunk://#diff-f97b528be07ea2a25c936948697e54a3fd1a2a92b02178888aae827b0334ee78R322)

**Documentation and README updates**
- Updates `README.md` to document the Blackboard API as a supported communication pattern, describes its use case, and adds instructions for running the new example. Also updates the project directory structure to reflect the new pattern. [[1]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R17) [[2]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L41-R43) [[3]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L80-R89) [[4]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R337-R348) [[5]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R389-R397)
- Updates `ARCHITECTURE.md` to list Blackboard as a supported pattern and makes minor table formatting improvements. [[1]](diffhunk://#diff-8f6366fd8e7a5fa30f7f05879999195b7cf5fa1e3d157822eb37f0e91d226cfdL68-R68) [[2]](diffhunk://#diff-8f6366fd8e7a5fa30f7f05879999195b7cf5fa1e3d157822eb37f0e91d226cfdL126-R126)
- Updates `CHANGELOG.md` to ensure consistent naming for Reactive Extensions.
- Improves formatting and clarity in the `examples/ArrayOfStructs/README.md`. [[1]](diffhunk://#diff-bf8d93427982f8ad0942f0358bfd6c956f8cfd9b6f04a1673b2b07c69bb7752bL14-R15) [[2]](diffhunk://#diff-bf8d93427982f8ad0942f0358bfd6c956f8cfd9b6f04a1673b2b07c69bb7752bR54) [[3]](diffhunk://#diff-bf8d93427982f8ad0942f0358bfd6c956f8cfd9b6f04a1673b2b07c69bb7752bL73-R75) [[4]](diffhunk://#diff-bf8d93427982f8ad0942f0358bfd6c956f8cfd9b6f04a1673b2b07c69bb7752bL96-R98) [[5]](diffhunk://#diff-bf8d93427982f8ad0942f0358bfd6c956f8cfd9b6f04a1673b2b07c69bb7752bL137-R145)

## Notes for Reviewer
<!-- Items in addition to the checklist below that the reviewer should look for -->

## Pre-Review Checklist for the PR Author

* [x] Add sensible notes for the reviewer
* [x] PR title is short, expressive and meaningful
* [x] Consider switching the PR to a draft (`Convert to draft`)
    * as draft PR, the CI will be skipped for pushes
* [x] Relevant issues are linked in the [References](#references) section
* [x] Branch follows the naming format (`csharp-iox2-123-short-description`)
* [x] Commits messages are according to this [guideline][commit-guidelines]
    * [ ] Commit messages have the issue ID (`[#123] Add feature description`)
* [x] Tests have been added/updated for new functionality
* [x] XML documentation comments added to public APIs
* [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
* [x] Assign PR to reviewer
* [x] All checks have passed

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[changelog]: https://github.com/eclipse-iceoryx/iceoryx2-csharp/blob/main/CHANGELOG.md

## PR Reviewer Reminders

* Commits are properly organized and messages are according to the guideline
* Unit tests have been written for new behavior
* Public API is documented with XML comments
* PR title describes the changes
* Code follows C# conventions (PascalCase for public APIs)
    * `dotnet format` has been exectued before submitting

## References

<!-- Use either 'Closes #123' or 'Relates to #123' to reference the corresponding issue. -->

Closes #4

patdhlk committed:
Add initial Blackboard support

patdhlk committed:
Merge branch 'eclipse-iceoryx:main' into csharp-iox2-4-blackboard-pattern-support

patdhlk committed:
Merge branch 'eclipse-iceoryx:main' into csharp-iox2-4-blackboard-pattern-support

dkroenke commented on examples/Blackboard/Blackboard.csproj:
Could you add a small Readme on how to run the example and what is happening there?
A link to https://ekxide.github.io/iceoryx2-book/main/fundamentals/messaging-patterns/blackboard.html would be help here.

patdhlk replied:
ok...will do in the future

patdhlk replied:
@dkroenke please wait with the review while it is still in `DRAFT` state

dkroenke reviewed (commented):
An entry in the Changelog would be good.

To keep the git history a bit more clean you could update the feature branches to latest main via `rebase` instead of `merge`. That avoids the merge commits in the history. In iceoryx2 we do it the same way.

dkroenke commented:
Adding @FerdinandSpitzschnueffler for the review of the Backboard components

FerdinandSpitzschnueffler commented on src/Iceoryx2/Blackboard/EntryHandle.cs:
```diff
@@ -0,0 +1,116 @@
+    /// Gets the current value from the blackboard entry.
+    /// </summary>
+    /// <returns>A Result containing the current value or an error.</returns>
+    public unsafe Result<TValue, Iox2Error> Get()
```
Do we need to return a `Result` here? Get shouldn't fail and it seems that we never return an `Err` in the function.

FerdinandSpitzschnueffler commented on src/Iceoryx2/Blackboard/EntryHandleMut.cs:
```diff
@@ -0,0 +1,159 @@
+    /// </summary>
+    /// <param name="value">The new value to set.</param>
+    /// <returns>A Result indicating success or an error.</returns>
+    public unsafe Result<Unit, Iox2Error> Update(TValue value)
```
With the last release, `update()` was renamed to `update_with_copy()`. Could you please rename this method as well for consistency?

FerdinandSpitzschnueffler replied:
Do we need to return a `Result`?

FerdinandSpitzschnueffler commented on src/Iceoryx2/Blackboard/EntryValue.cs:
```diff
@@ -0,0 +1,131 @@
+/// </summary>
+/// <typeparam name="TKey">The type of the key.</typeparam>
+/// <typeparam name="TValue">The type of the value.</typeparam>
+public sealed class EntryValue<TKey, TValue> : IDisposable
```
Could you please rename the class to `EntryValueUninit` for consistency?

FerdinandSpitzschnueffler commented on src/Iceoryx2/Blackboard/EntryValue.cs:
```diff
@@ -0,0 +1,131 @@
+    /// Gets a mutable reference to the payload for in-place modification.
+    /// </summary>
+    /// <returns>A reference to the payload.</returns>
+    public unsafe ref TValue PayloadMut()
```
Could you please rename the function to `ValueMut` for consistency?

FerdinandSpitzschnueffler commented on src/Iceoryx2/Blackboard/EntryValue.cs:
```diff
@@ -0,0 +1,131 @@
+    /// Writes a value to the loaned memory.
+    /// </summary>
+    /// <param name="value">The value to write.</param>
+    public unsafe void Write(TValue value)
```
With the last release, `write` and `update` have been combined to `update_with_copy`. Could you please adapt this here as well? FYI: https://github.com/eclipse-iceoryx/iceoryx2/blob/main/doc/release-notes/iceoryx2-v0.8.0.md#api-breaking-changes, item 9.

FerdinandSpitzschnueffler commented on src/Iceoryx2.Reactive/ReaderExtensions.cs:
```diff
@@ -0,0 +1,166 @@
+        {
+            var entryResult = reader.Entry<TValue>(key);
+
+            if (entryResult.IsOk)
```
Currently, blackboard entries cannot be removed, so it should be fine to move these lines outside the while loop.

FerdinandSpitzschnueffler commented on src/Iceoryx2.Reactive/ReaderExtensions.cs:
```diff
@@ -0,0 +1,166 @@
+
+            if (entryResult.IsOk)
+            {
+                using var entry = entryResult.Unwrap();
```
Same here.

patdhlk committed:
Add README for Blackboard example detailing usage and concepts

FerdinandSpitzschnueffler reviewed (commented):
I'm not familiar with C#, so I mainly focused on the blackboard API and not on the implementation. The PR looks good to me, but some function names have changed with the last release.

patdhlk committed:
Add enhanced Blackboard API support and README updates

- Introduced `BlackboardHelpers` for alignment handling.
- Refactored key alignment logic across Blackboard classes.
- Added thread safety remarks to `Writer`, `Reader`, and `BlackboardServiceBuilder`.
- Enhanced `BlackboardServiceBuilder` with input validation.
- Optimized memory usage in key-value operations via stack allocation.
- Updated README with examples and documentation for Blackboard API.
- Incorporated detailed comments and ensured proper resource disposal.

patdhlk committed:
Introduce `SafeEntryValueHandle` and enhance Blackboard pattern

- Added a new `SafeEntryValueHandle` to encapsulate native resource management safely.
- Improved `EntryValue` API for consistency and optimized memory handling.
- Reworked `Reader` and `Writer` logic to streamline key-value operations using `UpdateWithCopy`.
- Updated examples and documentation to reflect API changes and best practices.
- Refined markdown files with formatting fixes and clearer explanations.

patdhlk committed:
Fix formatting in `EntryValue` and `SafeEntryValueHandle`

FerdinandSpitzschnueffler commented on src/Iceoryx2/Blackboard/EntryValueUninit.cs:
The file could be renamed to `EntryValueUninit.cs`.

FerdinandSpitzschnueffler commented on src/Iceoryx2/Blackboard/EntryValue.cs:
```diff
@@ -0,0 +1,126 @@
+        var rawHandle = _handle.Consume();
+
+        // Note: iox2_entry_value_update returns void
+        iox2_entry_value_update(
```
I wonder why this works 🤔  Shouldn't this be `iox2_entry_value_uninit_update`?

patdhlk replied:
good catch...this does not work and would've cause a runtime crash

patdhlk committed:
[#4] Rename EntryValue to EntryValueUninit for FFI consistency
Address PR review findings by aligning naming with iceoryx2 FFI:
  - Rename EntryValue.cs to EntryValueUninit.cs
  - Rename SafeEntryValueHandle to SafeEntryValueUninitHandle
  - Update FFI method names to use _uninit suffix:
    - iox2_entry_value_mut -> iox2_entry_value_uninit_value_mut
    - iox2_entry_value_update -> iox2_entry_value_uninit_update
    - iox2_entry_value_drop -> iox2_entry_value_uninit_drop

dkroenke reviewed (approved):
Looks good, thanks for the contribution!

FerdinandSpitzschnueffler reviewed (approved):
Looks good and can be merged. Thank you for your contribution!
Out of curiosity: do you plan additional tests for the blackboard and the other messaging patterns?

patdhlk commented:
> Looks good and can be merged. Thank you for your contribution! Out of curiosity: do you plan additional tests for the blackboard and the other messaging patterns?

@FerdinandSpitzschnueffler yes, already in progress :)
